sbi/inference/potentials/score_based_potential.py

```diff
@@ -0,0 +1,330 @@
+
+        # Compute the base density
+        mean_T = self.score_estimator.mean_T
+        std_T = self.score_estimator.std_T
+        base_density = torch.distributions.Normal(mean_T, std_T)
+        # TODO: is this correct? should we use append base_density for each dimension?
+        for _ in range(len(self.score_estimator.input_shape)):
+            base_density = torch.distributions.Independent(base_density, 1)
+
+        # Build the freeform jacobian transformation by probability flow ODEs
+        transform = build_freeform_jacobian_transform(
+            self.score_estimator, condition, atol=atol, rtol=rtol, exact=exact
+        )
+        # Use zuko to build the normalizing flow.
+        return NormalizingFlow(transform, base=base_density)
+
+
+def build_freeform_jacobian_transform(
+    score_estimator,
+    x_o: Tensor,
+    atol: float = 1e-5,
+    rtol: float = 1e-6,
+    exact: bool = True,
+):
+    # Create a freeform jacobian transformation
+    phi = (x_o, *score_estimator.parameters())
+
+    def f(t, x):
+        score = score_estimator(input=x, condition=x_o, time=t)
+        f = score_estimator.drift_fn(x, t)
+        g = score_estimator.diffusion_fn(x, t)
+        v = f - 0.5 * g**2 * score
+        return v
+
+    transform = FreeFormJacobianTransform(
+        f=f,
+        t0=score_estimator.T_min,
+        t1=score_estimator.T_max,
+        phi=phi,
+        atol=atol,
+        rtol=rtol,
+        exact=exact,
+    )
+
+    return transform
+
+
+def _iid_bridge(
+    theta: Tensor,
+    xos: Tensor,
+    time: Tensor,
+    score_estimator: ConditionalScoreEstimator,
+    prior: Distribution,
+    t_max: float = 1.0,
+):
+    r"""
+    Returns the score-based potential for multiple IID observations. This can require a
+    special solver to obtain the correct tall posterior.
+
+    Args:
+        input: The parameter values at which to evaluate the potential.
+        condition: The observed data at which to evaluate the potential.
+        time: The diffusion time.
+        score_estimator: The neural network modelling the score.
+        prior: The prior distribution.
+    """
+
+    assert (
+        next(score_estimator.parameters()).device == xos.device
+        and xos.device == theta.device
+    ), f"""device mismatch: estimator, x, theta: \
+        {next(score_estimator.parameters()).device}, {xos.device},
+        {theta.device}."""
+
+    # Get number of observations which are left from event_shape if they exist.
+    condition_shape = score_estimator.condition_shape
+    num_obs = xos.shape[-len(condition_shape) - 1]
+
+    # Calculate likelihood in one batch.
+    # xos is of shape (num_obs, *condition_shape).
+    # theta is of shape (num_samples, *parameter_shape).
+
+    # we need to combine the batch shapes of num_obs and num_samples for both
+    # theta and xos.
+    # TODO: janfb adapted this to fix the iid setting shape problems, but iid inference
```

Comment:
Maybe we remove the IID code from main for the release? I think we said this is for a future PR.